# Introdoction

This is pipeline written in Python programming language. It is used for the analysis of *Mycobacterium tuberculosis* Methylation profile data generated using PacBio Sequencing system. The input data were generated by analysis of PacBio sequencing output using SMRT analysis tool.

# Objective 

	* Studying m6A Methylation profile
	* Phylogeny analysis of the provided sample
	* Selected gene analysis
	* SNPs analysis

# External Tools Require by Script
- [MUSCLE](https://www.drive5.com/muscle/)
- [RAxML](https://sco.h-its.org/exelixis/web/software/raxml/)
- [BLAT](http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/blat/)
- [MAFFT](https://mafft.cbrc.jp/alignment/software/)

# External  Python Modules  Required

All the package can be installed using pip command

- Biopython
- ruffus
- pandas
- ete3
- seaborn
- click

# Running Steps


# External data

- [ProteinTable](https://www.ncbi.nlm.nih.gov/genome/proteins/166?genome_assembly_id=159857)
- List of the genes need to be analysed


# Steps used before running the script

- Put all the gene sequences need to be analysed in separated file (*.ffn) in __selected_genes__ folder
- Download an outgroup genome. In our analysis [NC_015848](https://www.ncbi.nlm.nih.gov/genome/?term=NC_015848) is used.
- Protein table was uploaded to [Some Server]() to indetify the operons. Operons file was downloaded and saved as ......
- Outgroup genome was selected in the analysis NC_015848.1.fna

# File Arrangements

Each sample must have two Folders, 5_BASE_MODIFICATIONS and Analysis.5_BASE_MODIFICATIONS should contain modifications.gff.gz and Analysis folder should contain consensus.fasta

# Contributions

# System tested

Ubuntu 16.04 64bit

